### PR TITLE
Allow automatic acquisition of n GPUs through --device-ids -n

### PR DESCRIPTION
**Pull request by fhieber**
More flexible way of automatically acquiring multiple GPUs.

```--device-ids -4``` is equivalent to ```--device-ids -1 -2 -1```

### CODE REVIEW

File: docs/user_documentation.md [hunk 1]
```diff
@@ -78,20 +78,22 @@ Then start tensorboard and point it to the model directory (or any parent direct
 ### CPU/GPU training
 
 By default, training is carried out on the first GPU device of your machine.
-You can specify alternative devices with the `--device-ids` option, with
+You can specify alternative GPU devices with the `--device-ids` option, with
 which you can also activate multi-GPU training (see below). If
 `--device-ids -1`, sockeye will try to find a free GPU on your machine and block
-until one is available. The locking mechanism is based on files and therefore assumes all processes are running
+until all of them are available. The locking mechanism is based on files and therefore assumes all processes are running
```
Comment (davvil) [inline docs/user_documentation.md]: Unclear, specially the "all of them" as in the example you set it to -1. I would write something along the lines of "If you specify a negative number, sockeye will try to allocate this number of given GPUs, blocking until they become available."

File: sockeye/arguments.py [hunk 2]
```diff
@@ -85,9 +85,11 @@ def add_device_args(params):
     device_params = params.add_argument_group("Device parameters")
 
     device_params.add_argument('--device-ids', default=[-1],
-                               help='List of GPU ids to use. Default: %(default)s. '
-                                    'Use -1 to automatically acquire a GPU through a file locking mechanism. '
-                                    '(Note that this assumes GPU processes are using automatic sockeye GPU ids).',
+                               help='List or number of GPUs ids to use. Default: %(default)s. '
+                                    'Use -x to automatically acquire x GPUs through a file locking mechanism. '
```
Comment (davvil) [inline sockeye/arguments.py]: I would remove the "file locking mechanism".

**Comment by davvil** [comment]
Additional comment: In your commit message you probably wanted "--device-ids -1 -1 -1" (not -2)

**Comment by fhieber** [comment]
Actually, -1 -2 -1 was intended as this would automatically acquire 4 gpus, equivalent to the -4

File: docs/user_documentation.md (hunk 1 above)
Comment (fhieber) [inline docs/user_documentation.md]: Thanks, I forgot to go back to the original version in this part of the sentence. I initially had something similar to what you suggest, but the documentation first covers single-device training and then turns to multi-device. Will fix the sentence.

File: sockeye/arguments.py (hunk 2 above)
Comment (fhieber) [inline sockeye/arguments.py]: yeah right, its not really making it clearer there.

File: sockeye/arguments.py [hunk 3]
```diff
@@ -85,9 +85,11 @@ def add_device_args(params):
     device_params = params.add_argument_group("Device parameters")
 
     device_params.add_argument('--device-ids', default=[-1],
-                               help='List of GPU ids to use. Default: %(default)s. '
-                                    'Use -1 to automatically acquire a GPU through a file locking mechanism. '
-                                    '(Note that this assumes GPU processes are using automatic sockeye GPU ids).',
+                               help='List or number of GPUs ids to use. Default: %(default)s. '
+                                    'Use -x to automatically acquire x GPUs. '
+                                    'Use x [x] to use a specific GPU id on this host. '
```
Comment (tdomhan) [inline sockeye/arguments.py]: what's 'x [x]' supposed to mean?

Comment (fhieber) [inline sockeye/arguments.py]: at least one id, optionally followed by another :)

**Pull request merged**